text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,130 @@
+# OpenTelemetry versioning and stability
+
+OpenTelemetry is a large project with strict compatibility requirements. This proposal defines the stability guarantees offered by OpenTelemetry, along with a versioning and lifecycle proposal which defines how we meet those requirements.
+
+Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.
```

Comment:
OpenTelemetry does use RFC 2119, but this is an OTEP which can have a more vague wording. The expectation is that once the OTEP is approved, a new document with more precise wording where necessary is added to the spec based on it.

Reply:
Yes, correct. Once we have feedback from the language specific proposals currently being drafted, it will be clearer what can feasibly be listed as a hard requirement when we add this to the spec. 

But, I should emphasize that the ultimate goal of this OTEP is not strict compliance with a specific implementation approach. It is far more important that implementors understand the stability requirements, and meet those requirements. Achieving backwards compatibility is an area where programming languages differ greatly, in terms of language requirements and mechanisms. Some languages have ABI requirements; other languages don't even have types. So I expect some amount of divergence in the implementation details, but no divergence in the strict stability and compatibility guarantees we want to offer. 

If the stability and depreciation goals are not clear, please let me know so we can clarify further.

Reply:
> OpenTelemetry does use RFC 2119, but this is an OTEP (proposal) which can have a more vague wording. The expectation is that once the OTEP is approved, a new document with more precise wording where necessary is added to the spec based on it.

@Oberon00 Hm... probably you meant "OpenTelemetry spec", not "OpenTelemetry", right? (I'm still struggling to figure out terms in this project :wink: )

Since this OTEP addresses recommendation/expectation/requirement, and the level of requirement is the core of this OTEP... I believe it should be more clear than now - although there are much more room in other types of OTEP. I don't think it's a good idea to delay "precise term" discussion to PR in the spec.

> I expect some amount of divergence in the implementation details, but no divergence in the strict stability and compatibility guarantees we want to offer.

@tedsuo I like the distinction indeed! However what are "strict" or "hard" requirement is not clear unless we use clear terms to describe them. That's why we have to use RFC 2119 even in this OTEP to make it clear.

Reply:
> Hm... probably you meant "OpenTelemetry spec"

OTEPs are expected to be very general informal proposals (after one OTEP is accepted _then_ it has to be incorporated into the actual Specification, which is when it has to have the proper terms, etc). I suggest don't getting lost into the details now (we will have time for that after this part hits the Spec).